Mathlib/RingTheory/PiTensorProduct.lean

```diff
@@ -0,0 +1,295 @@
+
+lemma zero_lmul (x : ⨂[R] i, A i) : lmul 0 x = 0 := by
+  induction' x using PiTensorProduct.induction_on <;> simp
+
+lemma lmul_zero (x : ⨂[R] i, A i) : lmul x 0 = 0 := by
+  induction' x using PiTensorProduct.induction_on <;> simp
+
+lemma lmul_add (x y z : ⨂[R] i, A i) : lmul x (y + z) = lmul x y + lmul x z := by
+  induction' x using PiTensorProduct.induction_on <;> simp
+
+lemma add_lmul (x y z : ⨂[R] i, A i) : lmul (x + y) z = lmul x z + lmul y z := by
+  induction' x using PiTensorProduct.induction_on <;> simp
+
+instance nonUnitalNonAssocSemiring : NonUnitalNonAssocSemiring (⨂[R] i, A i) where
+  __ := mul
+  __ := inferInstanceAs (AddCommMonoid (⨂[R] i, A i))
+  left_distrib := lmul_add
+  right_distrib := add_lmul
+  zero_mul := zero_lmul
+  mul_zero := lmul_zero
+
+end NonUnitalNonAssocSemiring
+
+noncomputable section NonAssocSemiring
+
+variable [CommSemiring R] [∀ i, NonAssocSemiring (A i)]
+variable [∀ i, Module R (A i)] [∀ i, SMulCommClass R (A i) (A i)] [∀ i, IsScalarTower R (A i) (A i)]
+
+lemma one_lmul (x : ⨂[R] i, A i) : lmul (tprod R 1) x = x := by
+  induction' x using PiTensorProduct.induction_on with rx x x₁ x₂ hx₁ hx₂
+  · simp
+  · simp only [map_add, hx₁, hx₂]
+
+lemma lmul_one (x : ⨂[R] i, A i) : lmul x (tprod R 1) = x := by
+  induction' x using PiTensorProduct.induction_on with rx x x₁ x₂ hx₁ hx₂
+  · simp
+  · simp only [map_add, LinearMap.add_apply, hx₁, hx₂]
+
+instance nonAssocSemiring : NonAssocSemiring (⨂[R] i, A i) where
+  __ := nonUnitalNonAssocSemiring
+  one_mul := one_lmul
+  mul_one := lmul_one
+
+end NonAssocSemiring
+
+noncomputable section NonUnitalSemiring
+
+variable [CommSemiring R] [∀ i, NonUnitalSemiring (A i)]
+variable [∀ i, Module R (A i)] [∀ i, SMulCommClass R (A i) (A i)] [∀ i, IsScalarTower R (A i) (A i)]
+
+lemma lmul_assoc (x y z : ⨂[R] i, A i) : lmul (lmul x y) z = lmul x (lmul y z) := by
+  induction' x using PiTensorProduct.induction_on with rx x x₁ x₂ hx₁ hx₂
+  · induction' y using PiTensorProduct.induction_on with ry y y₁ y₂ hy₁ hy₂
+    · induction' z using PiTensorProduct.induction_on with rz z z₁ z₂ hz₁ hz₂
+      · simp only [map_smul, LinearMap.smul_apply, lmul_tprod_tprod, mul_assoc]
+      · simp only [map_smul, LinearMap.smul_apply, lmul_tprod_tprod, map_add] at hz₁ hz₂ ⊢
+        rw [hz₁, hz₂]
+    · simp only [map_smul, LinearMap.smul_apply, map_add, LinearMap.add_apply] at hy₁ hy₂ ⊢
+      rw [hy₁, hy₂]
+  · simp only [map_add, LinearMap.add_apply] at hx₁ hx₂ ⊢
+    rw [hx₁, hx₂]
+
+instance nonUnitalSemiring : NonUnitalSemiring (⨂[R] i, A i) where
+  __ := nonUnitalNonAssocSemiring
+  mul_assoc := lmul_assoc
+
+end NonUnitalSemiring
+
+noncomputable section Semiring
+
+variable [CommSemiring R] [∀ i, Semiring (A i)] [∀ i, Algebra R (A i)]
+
+instance semiring : Semiring (⨂[R] i, A i) where
+  __ := nonUnitalSemiring
+  __ := nonAssocSemiring
+
+instance algebra : Algebra R (⨂[R] i, A i) where
+  __ := hasSMul'
+  toFun := (· • 1)
```

Comment:
Can you add a lemma that say this is equal to `tprod _ (Pi.mulSingle i (algebraMap _ _ r))`?